homeassistant/components/plum_lightpad/__init__.py

```diff
@@ -26,56 +27,51 @@
     extra=vol.ALLOW_EXTRA,
 )
 
+PLATFORMS = ["light"]
 
-async def async_setup(hass, config):
+
+async def async_setup(hass: HomeAssistant, config: dict):
     """Plum Lightpad Platform initialization."""
+    if DOMAIN not in config:
+        return True
 
     conf = config[DOMAIN]
-    plum = Plum(conf[CONF_USERNAME], conf[CONF_PASSWORD])
 
-    hass.data[DOMAIN] = plum
+    _LOGGER.info("Found Plum Lightpad configuration in config, importing...")
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
+        )
+    )
 
-    def cleanup(event):
-        """Clean up resources."""
-        plum.cleanup()
+    return True
 
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, cleanup)
 
-    cloud_web_sesison = async_get_clientsession(hass, verify_ssl=True)
-    await plum.loadCloudData(cloud_web_sesison)
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Plum Lightpad from a config entry."""
+    _LOGGER.debug("Setting up config entry with ID = %s", entry.unique_id)
 
-    async def new_load(device):
-        """Load light and sensor platforms when LogicalLoad is detected."""
-        await asyncio.wait(
-            [
-                hass.async_create_task(
-                    discovery.async_load_platform(
-                        hass, "light", DOMAIN, discovered=device, hass_config=conf
-                    )
-                )
-            ]
-        )
+    username = entry.data.get(CONF_USERNAME)
+    password = entry.data.get(CONF_PASSWORD)
 
-    async def new_lightpad(device):
-        """Load light and binary sensor platforms when Lightpad detected."""
-        await asyncio.wait(
-            [
-                hass.async_create_task(
-                    discovery.async_load_platform(
-                        hass, "light", DOMAIN, discovered=device, hass_config=conf
-                    )
-                )
-            ]
-        )
+    try:
+        plum = await load_plum(username, password, hass)
 
-    device_web_session = async_get_clientsession(hass, verify_ssl=False)
-    hass.async_create_task(
-        plum.discover(
-            hass.loop,
-            loadListener=new_load,
-            lightpadListener=new_lightpad,
-            websession=device_web_session,
-        )
-    )
+        hass.data.setdefault(DOMAIN, {})
+        hass.data[DOMAIN][entry.entry_id] = plum
 
-    return True
+        for component in PLATFORMS:
+            hass.async_create_task(
+                hass.config_entries.async_forward_entry_setup(entry, component)
+            )
+
+        def cleanup(event):
+            """Clean up resources."""
+            plum.cleanup()
+
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, cleanup)
+        return True
+
+    except (ContentTypeError, ConnectTimeout, HTTPError) as ex:
```

Comment:
We should only wrap the code in a `try…except` that can raise these errors

Reply:
Can we write a test for these errors?

Reply:
makes sense, done